src/shell_utils.cpp

```diff
@@ -0,0 +1,169 @@
+#include "shell_utils.h"
+#include <sstream>
+#include <iostream>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+// ---- c-tor for struct Job -----
+Job::Job(const std::string& cmd, int processId, bool running)
+    : command(cmd), pid(processId), isRunning(running) {
+}
+
+std::vector<Job> backgroundJobs; // Vector to store background jobs
+
+// ----- Function definitions -----
+// Splits a string into a vector of tokens based on a delimiter
+std::vector<std::string> split(const std::string& input, char delimiter) {
+    std::vector<std::string> tokens;
+    std::string token;
+    std::istringstream tokenStream(input);
+    while (std::getline(tokenStream, token, delimiter)) {
+        tokens.push_back(token);
+    }
+    return tokens;
+}
+
+// Executes a child process with the given command path and arguments
+void executeChildProcess(const std::string& commandPath, const std::vector<std::string>& arguments) {
+    std::vector<char*> args;
+    args.push_back(const_cast<char*>(commandPath.c_str())); // First argument is the command itself
+
+    // Prepare arguments for execvp
+    for (const auto& arg : arguments) {
+        args.push_back(const_cast<char*>(arg.c_str()));
+    }
+    args.push_back(nullptr); // Last argument must be nullptr
+
+    // Execute the command
+    if (execvp(commandPath.c_str(), args.data()) == -1) {
+        std::cerr << "Failed to execute command." << std::endl;
+        exit(EXIT_FAILURE);	// In case execvp fails
+    }
+}
+
+// Executes a command with the given command path, arguments, and background execution flag
+void executeCommand(const std::string& commandPath, const std::vector<std::string>& arguments, bool runInBackground) {
+    pid_t pid = fork();
+    if (pid < 0) {
+        std::cerr << "Fork failed." << std::endl;
+        return;
+    }
+    if (pid == 0) {
+        // Child process
+        executeChildProcess(commandPath, arguments);
+    }
+    else {
+        // Parent process
+        if (!runInBackground) {
+            int status;
+            waitpid(pid, &status, 0); // Wait for the child process to complete
+            // Update the job status based on the process exit status
+            bool isRunning = !WIFEXITED(status) || (WEXITSTATUS(status) == 0);
+            updateJobStatus(pid, isRunning);
+        }
+        else {
+            // Add the child process details (pid, command, etc.) to a list of background jobs
+            Job job(commandPath, pid, true);
+            backgroundJobs.push_back(job);
```

Comment:
To avoid copying the `Job` struct into the vector, you can use `std::move`. 
Also, you can use in place construction which has the best performance in this case. Check it out (where it talks about the `emplace_back` method): https://arne-mertz.de/2016/02/modern-c-features-in-place-construction/